core/lib/basic_types/src/protocol_version.rs

```diff
@@ -255,3 +278,103 @@ impl From<ProtocolVersionId> for VmVersion {
         }
     }
 }
+
+basic_type!(
+    /// Patch part of semantic protocol version.
+    VkPatch,
+    u32
+);
+
+/// Semantic protocol version.
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, SerializeDisplay, DeserializeFromStr, Hash, PartialOrd, Ord,
+)]
+pub struct ProtocolSemanticVersion {
+    pub minor: ProtocolVersionId,
+    pub patch: VkPatch,
+}
+
+impl ProtocolSemanticVersion {
+    const MAJOR_VERSION: u8 = 0;
+
+    pub fn try_from_packed(packed: U256) -> Result<Self, String> {
+        let minor = ((packed >> U256::from(PACKED_SEMVER_MINOR_OFFSET))
+            & U256::from(PACKED_SEMVER_MINOR_MASK))
+        .try_into()?;
+        let patch = packed.0[0] as u32;
+
+        Ok(Self {
+            minor,
+            patch: VkPatch(patch),
+        })
+    }
+
+    pub fn pack(&self) -> U256 {
+        (U256::from(self.minor as u16) << U256::from(PACKED_SEMVER_MINOR_OFFSET))
+            | U256::from(self.patch.0)
+    }
+}
+
+impl fmt::Display for ProtocolSemanticVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}.{}.{}",
+            Self::MAJOR_VERSION,
+            self.minor as u16,
+            self.patch
+        )
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ParseProtocolSemanticVersionError {
+    #[error("invalid format")]
+    InvalidFormat,
+    #[error("non zero major version")]
+    NonZeroMajorVersion,
+    #[error("{0}")]
+    ParseIntError(ParseIntError),
+}
+
+impl FromStr for ProtocolSemanticVersion {
+    type Err = ParseProtocolSemanticVersionError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<&str> = s.split('.').collect();
+        if parts.len() != 3 {
+            return Err(ParseProtocolSemanticVersionError::InvalidFormat);
+        }
+
+        let major = parts[0]
+            .parse::<u16>()
+            .map_err(ParseProtocolSemanticVersionError::ParseIntError)?;
+        if major != 0 {
+            return Err(ParseProtocolSemanticVersionError::NonZeroMajorVersion);
+        }
+
+        let minor = parts[1]
+            .parse::<u16>()
+            .map_err(ParseProtocolSemanticVersionError::ParseIntError)?;
+        let minor = ProtocolVersionId::try_from(minor)
+            .map_err(|_| ParseProtocolSemanticVersionError::InvalidFormat)?;
+
+        let patch = parts[2]
+            .parse::<u32>()
```

Comment:
I wonder if we also should not allow 'minor' to be 0..

as this generally means that we're reading data from before the semver upgrade

Reply:
This method is only used for parsing newly added config variable. It won't read from old variables and they have different format anyway (not a semver but just a number)

Reply:
ok